### ISSUE TITLE
[Bug] System.TypeLoadException: Could not load type 'Microsoft.IdentityModel.Json.JsonConvert' from assembly 'Microsoft.IdentityModel.Tokens, Version=7.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'.

### ISSUE THREAD

**Issue by ManuelHaas** [AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet#2346]
**Which version of Microsoft.IdentityModel are you using?**
Microsoft.IdentityModel 7.0.0 & 7.0.2

**Where is the issue?**
 *  M.IM.Tokens
 *  M.IM.Tokens.Saml

**Is this a new or an existing app?**
The app is in production and I have upgraded Microsoft.IdentityModel.Tokens.Saml from Version="6.32.1"  to Version="7.0.0".
After upgrade to 7.0.2 the error still persists.

**Repro**
Code leads to the following exception after update:
```csharp
2023-10-03 15:29:20.624 +02:00 [ERR] An unhandled exception has occurred while executing the request.
System.InvalidOperationException: IDX20803: Unable to obtain configuration from: 'https://login.microsoftonline.com/XYZ/v2.0/.well-known/openid-configuration'. Will retry at '03.10.2023 13:29:20 +00:00'. Exception: 'System.TypeLoadException: Could not load type 'Microsoft.IdentityModel.Json.JsonConvert' from assembly 'Microsoft.IdentityModel.Tokens, Version=7.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'.
   at Microsoft.IdentityModel.Protocols.OpenIdConnect.OpenIdConnectConfigurationRetriever.GetAsync(String address, IDocumentRetriever retriever, CancellationToken cancel)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.IdentityModel.Protocols.OpenIdConnect.OpenIdConnectConfigurationRetriever.GetAsync(String address, IDocumentRetriever retriever, CancellationToken cancel)
   at Microsoft.IdentityModel.Protocols.OpenIdConnect.OpenIdConnectConfigurationRetriever.Microsoft.IdentityModel.Protocols.IConfigurationRetriever<Microsoft.IdentityModel.Protocols.OpenIdConnect.OpenIdConnectConfiguration>.GetConfigurationAsync(String address, IDocumentRetriever retriever, CancellationToken cancel)
   at Microsoft.IdentityModel.Protocols.ConfigurationManager`1.GetConfigurationAsync(CancellationToken cancel)'.
 ---> System.TypeLoadException: Could not load type 'Microsoft.IdentityModel.Json.JsonConvert' from assembly 'Microsoft.IdentityModel.Tokens, Version=7.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'.
   at Microsoft.IdentityModel.Protocols.OpenIdConnect.OpenIdConnectConfigurationRetriever.GetAsync(String address, IDocumentRetriever retriever, CancellationToken cancel)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.IdentityModel.Protocols.OpenIdConnect.OpenIdConnectConfigurationRetriever.GetAsync(String address, IDocumentRetriever retriever, CancellationToken cancel)
   at Microsoft.IdentityModel.Protocols.OpenIdConnect.OpenIdConnectConfigurationRetriever.Microsoft.IdentityModel.Protocols.IConfigurationRetriever<Microsoft.IdentityModel.Protocols.OpenIdConnect.OpenIdConnectConfiguration>.GetConfigurationAsync(String address, IDocumentRetriever retriever, CancellationToken cancel)
   at Microsoft.IdentityModel.Protocols.ConfigurationManager`1.GetConfigurationAsync(CancellationToken cancel)
   --- End of inner exception stack trace ---
   at Microsoft.IdentityModel.Protocols.ConfigurationManager`1.GetConfigurationAsync(CancellationToken cancel)
   at Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectHandler.HandleChallengeAsyncInternal(AuthenticationProperties properties)
   at Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectHandler.HandleChallengeAsync(AuthenticationProperties properties)
   at Microsoft.AspNetCore.Authentication.AuthenticationHandler`1.ChallengeAsync(AuthenticationProperties properties)
   at Microsoft.AspNetCore.Authentication.AuthenticationService.ChallengeAsync(HttpContext context, String scheme, AuthenticationProperties properties)
   at Microsoft.AspNetCore.Mvc.ChallengeResult.ExecuteResultAsync(ActionContext context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResultFilterAsync>g__Awaited|30_0[TFilter,TFilterAsync](ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResultExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.ResultNext[TFilter,TFilterAsync](State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeResultFilters()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Serilog.AspNetCore.RequestLoggingMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddlewareImpl.<Invoke>g__Awaited|8_0(ExceptionHandlerMiddlewareImpl middleware, HttpContext context, Task task)
```

Error probably in line 8:
```csharp
 ---> System.TypeLoadException: Could not load type 'Microsoft.IdentityModel.Json.JsonConvert' from assembly 'Microsoft.IdentityModel.Tokens, Version=7.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'.
```

**Comment by christophe-mommer** [comment]
Same for me, it blows with the following code:

```
 var claims = new List<Claim>
 {
     new(JwtRegisteredClaimNames.Sub, user.Id),
     new(JwtRegisteredClaimNames.Email, user.UserName!)
 };

 var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(IdentityConsts.TokenKey));
 var creds = new SigningCredentials(key, SecurityAlgorithms.HmacSha256);
 var expires = DateTime.UtcNow.AddHours(1);

 var jwtSecurityTokenHandler = new JwtSecurityTokenHandler();
 var token = jwtSecurityTokenHandler.CreateJwtSecurityToken(
     new SecurityTokenDescriptor
     {
         Subject = new ClaimsIdentity(claims),
         Issuer = IdentityConsts.Issuer,
         Audience = IdentityConsts.Audience,
         Expires = expires,
         SigningCredentials = creds
     }
 ); // throw here
return token.ToString();
```
Note: I'm running in .NET 8 RC1

**Comment by jennyf19** [comment]
Are all your M.IM versions the same? https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/wiki/IdentityModel-7x#reminder

**Comment by jmprieur** [comment]
See https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/wiki/IdentityModel-7x#reminder

**Comment by ManuelHaas** [comment]
@jennyf19 @jmprieur 
I have only one reference to an IdentityModel library in my solution:
`<PackageReference Include="Microsoft.IdentityModel.Tokens.Saml" Version="6.33.0" />
`
And it works if it is not updated to 7.0.0 or 7.0.2.
According to NuGet all M$ packages except the above one and Microsoft.TypeScript.MSBuild are up to date.

**Comment by brentschmaltz** [comment]
@ManuelHaas Microsoft.IdentityModel.Json.JsonConvert is no longer a type in M$ packages.
This has all the markings of mixed assemblies 6x and 7x.

**Comment by ManuelHaas** [comment]
@brentschmaltz Ok, thanks for the info. That's what the error message suggests. And what does that mean now?

**Comment by springy76** [comment]
give this one a try:
https://stackoverflow.com/questions/77098132/error-generating-jwt-token-during-asp-net-core-web-api-authorization/77133355#comment136008891_77133355

**Comment by brentschmaltz** [comment]
@ManuelHaas make sure all the IdentityModel assemblies have the same version.

**Comment by springy76** [comment]
Please someone explain/reason this bullshit:
https://www.nuget.org/packages/Microsoft.AspNetCore.Authentication.OpenIdConnect/7.0.12#dependencies-body-tab

**Comment by stojancho-jefremov** [comment]
> @ManuelHaas make sure all the IdentityModel assemblies have the same version.

For me the solution was this. In our project we started seeing similar error as reported here, as soon as we bumped our package **System.IdentityModel.Tokens.Jwt** from **6.32.3** to **7.0.3**. After doing thorough investigation and seeing the comments from @jmprieur and @brentschmaltz in this issue I noticed that in our project we reference **IdentityModel.AspNetCore 4.3.0** which transitively referenced **Microsoft.IdentityModel.Protocols.OpenIdConnect 6.15.1**. As soon as I added direct reference for **Microsoft.IdentityModel.Protocols.OpenIdConnect 7.0.3** (the same version as other **IdentityModel** libraries in our project) the problem went away.

**Comment by brentschmaltz** [comment]
@stojancho-jefremov apologies for you hitting this error, we have an outstanding feature to check and throw (or warn) that the IdentityModel assemblies are not all the same version.

We, as a team, have become aware of the issue with removing or changing internal methods, for non major releases which lead to these issues.
In this case, we dropped our internal Json clone and bumped the major, but in the past we have removed / changed internals during non major releases.
